### PR TITLE
Automatically cleanup empty dirs in LocalFileSystem

### LINKED ISSUE #5976: Cleanup Empty Directories in LocalFilesystem
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently the way to delete a "directory" through the `ObjectStore` interface is to call `list` with the directory path as prefix and then pipe that to `delete_stream`. 

This is the correct way to do this for object storage since "directories" don't really exist and the path is just a prefix to individual objects. 

But for actual filesystems directories are actual inodes themselves in the filesystem which should be cleaned up. 

**Describe the solution you'd like**

Add a new method to `ObjectStore`: `async fn delete_prefix(&self, prefix: Option<&Path>) -> Result<()>` 

This default implementation for this method can just be calling `list(prefix)` and then piping that to `delete_stream` but in the case of `LocalFilesystem` (and potentially other implementations in the future) will be specialized to also delete directories 

**Describe alternatives you've considered**

This could be handled inside `LocalFilesystem` during `delete_stream` but could become complicated. The implementation would have to:
1. Track prefixes during the delete operation
2. After delete determine if any of the prefixes are now empty
3. Delete them

**Additional context**

### PR DESCRIPTION
**Pull request by fsdvh**
# Which issue does this PR close?

Closes #5976 
Closes https://github.com/apache/arrow-rs/pull/5975

# Rationale for this change
 
We want to automatically clean empty directories

# What changes are included in this PR?

Update for delete method implementation for `LocalFileSystem`

# Are there any user-facing changes?

Yes

### CODE REVIEW

**Comment by tustvold** [comment]
Can we please make this configurable, disabled by default, and add some test coverage

File: object_store/src/local.rs [hunk 1]
```diff
@@ -465,13 +465,31 @@ impl ObjectStore for LocalFileSystem {
...
-                _ => Error::UnableToDeleteFile { path, source: e }.into(),
-            }),
+        maybe_spawn_blocking(move || {
+            if let Err(e) = std::fs::remove_file(&path) {
+                Err(match e.kind() {
+                    ErrorKind::NotFound => Error::NotFound { path, source: e }.into(),
+                    _ => Error::UnableToDeleteFile { path, source: e }.into(),
+                })
+            } else {
+                let root = &config.root;
```
Comment (alamb) [inline object_store/src/local.rs:477]: Maybe we could add a comment here explaining what is going on / the rationale for this (to clean out directories that were now empty)

File: object_store/src/local.rs [hunk 2]
```diff
@@ -465,13 +465,31 @@ impl ObjectStore for LocalFileSystem {
...
+                let root = root
+                    .to_file_path()
+                    .map_err(|_| Error::InvalidUrl { url: root.clone() })?;
+                let mut parent = path.parent();
+
+                while let Some(loc) = parent {
```
Comment (alamb) [inline object_store/src/local.rs:483]: This looks like a good plan to me -- walk up the path trying to remove any empty directories found, stopping on first error or root

**Review by alamb — COMMENTED** [review]
Thank you @fsdvh  -- this looks like a nice implementation to me

Can you please add a test for this functionality (so we don't introduce regressions in the future by accident)?

cc @thinkharderdev

**Comment by fsdvh** [comment]
@alamb added a test to catch an expected behavior and a few comments

**Review by thinkharderdev — APPROVED** [review]
nice @fsdvh

**Comment by alamb** [comment]
Thanks again @fsdvh and @thinkharderdev

**Pull request merged**